src/engine/segment.go

```diff
@@ -431,3 +443,86 @@ func (segment *Segment) SetText() {
 		segment.text = strings.NewReplacer("`", "\\`", `%`, `%%`).Replace(segment.text)
 	}
 }
+
+// userSegments is a list of segment writers that have been added at runtime.
+var userSegments = make(map[SegmentType]SegmentWriter)
```

Comment:
I would only initialise this when we are about to use it, so validate this in `AddSegment` and initialise if needed.